gui/src/pages/NewProjectConfig.svelte

```diff
@@ -0,0 +1,156 @@
+<script lang="ts">
+  import Layout from '../components/Layout.svelte';
+  import Textbox from '../components/Textbox.svelte';
+  import Spinner from '../components/Spinner.svelte';
+  import ErrorLabel from '../components/ErrorLabel.svelte';
+  import SubmitButton from '../components/form/SubmitButton.svelte';
+  import CenterFormPanel from '../components/form/CenterFormPanel.svelte';
+  import type { ReadDirResult } from '../lib/api';
+  import { api, isClientError } from '../lib/api';
+  import * as router from 'svelte-spa-router';
+
+  export let params = { starter: '' };
+
+  const { starter } = params;
+
+  let name = starter;
+  let templateUrl: string | null = null;
+  let error: Error | null = null;
+
+  let workingDirectory: string | null = null;
+
+  const setWorkingDirectory = (dir: string) => {
+    workingDirectory = dir;
+    dirPromise = api.kernel.readDir(workingDirectory);
+  };
+
+  let dirPromise: Promise<ReadDirResult> = api.kernel
+    .getUserHomeDir()
+    .then((dir) => {
+      setWorkingDirectory(dir);
+      return dirPromise;
+    });
+</script>
+
+<Layout>
+  <CenterFormPanel title={`New ${starter} project`} backRoute="#/new-project">
+    <form
+      on:submit|preventDefault={async () => {
+        error = null;
+        let workspaceId;
+
+        if (starter !== 'empty') {
+          templateUrl = String(
+            (await api.kernel.describeTemplates()).find(
+              (x) => x.name === starter,
+            )?.url,
+          );
+        }
+
+        try {
+          workspaceId = await api.kernel.createProject(
+            `${workingDirectory}/${name}`,
+            templateUrl,
+          );
+          router.push(`/workspaces/${encodeURIComponent(workspaceId)}`);
+        } catch (ex) {
+          if (!isClientError(ex)) {
+            throw ex;
+          }
+          error = ex;
+        }
+      }}
+    >
+      <h1>New {starter} project</h1>
+
+      <label for="name">Name:</label>
+      <Textbox
+        bind:value={name}
+        name="name"
+        id="name"
+        placeholder={starter}
+        --input-width="100%"
+      />
+      <div style="height:32px" />
+      {#if workingDirectory}
+        <label for="root">Root:</label>
+        <h2><span>{workingDirectory}<span>/{name}</span></span></h2>
+        {#await dirPromise}
+          <Spinner />
+        {:then dir}
+          {#if dir.parent !== null}
+            <button
+              on:click={() => setWorkingDirectory(String(dir.parent?.path))}
```

Comment:
Checking whether `dir.parent` is nil would be both clearer and more correct. Here you introduce a race condition where if the user clicks on the button and then `dir` is changed and then the `on:click` handler is fired for the button press, `workingDirectory` will be set to `"undefined"`. (Obviously the chances of that happening are virtually zero — but it seems needless.)

Reply:
I'm not sure what you mean here @BenElgar - isn't my `{#if ...}` block doing this already?

Reply:
`dir.parent`, apparently, should be `string | null`, so the only reason I needed `String()` and `?.` was because VS Code doesn't seem to understand the type information that `{:then dir} {#if dir.parent !== null}` implies

Reply:
No — you're in an event handler which is a different context.

Let's say you've got a very slow computer (or a very busy one) and you're on this page. `dir.parent` is set so you can see the button. You click on the button. Now, because your computer is so slow, the event handler takes several hundred milliseconds to fire. In that time you click the button again, since it's still there since the event handler that would remove it hasn't fired yet. Now the first event handler runs, correctly navigating to the root directory and removing the button. Then the second one runs and sets your working directory to `"undefined"`, breaking the page.